File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/ObjectCreationExpressionSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class ObjectCreationExpressionSyntax : NewExpressionSyntax
	{
		internal TypeSyntax _type;

		internal ArgumentListSyntax _argumentList;

		internal ObjectCreationInitializerSyntax _initializer;

		public new SyntaxToken NewKeyword => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationExpressionSyntax)base.Green)._newKeyword, base.Position, 0);

		public new SyntaxList<AttributeListSyntax> AttributeLists
		{
			get
			{
				SyntaxNode red = GetRed(ref _attributeLists, 1);
				return new SyntaxList<AttributeListSyntax>(red);
			}
		}

		public TypeSyntax Type => GetRed(ref _type, 2);

		public ArgumentListSyntax ArgumentList => GetRed(ref _argumentList, 3);

		public ObjectCreationInitializerSyntax Initializer => GetRed(ref _initializer, 4);

		internal ObjectCreationExpressionSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal ObjectCreationExpressionSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax newKeyword, SyntaxNode attributeLists, TypeSyntax type, ArgumentListSyntax argumentList, ObjectCreationInitializerSyntax initializer)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationExpressionSyntax(kind, errors, annotations, newKeyword, attributeLists?.Green, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeSyntax)type.Green, (argumentList != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArgumentListSyntax)argumentList.Green) : null, (initializer != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationInitializerSyntax)initializer.Green) : null), null, 0)
		{
		}

		internal override SyntaxToken GetNewKeywordCore()
		{
			return NewKeyword;
		}

		internal override NewExpressionSyntax WithNewKeywordCore(SyntaxToken newKeyword)
		{
			return WithNewKeyword(newKeyword);
		}

		public new ObjectCreationExpressionSyntax WithNewKeyword(SyntaxToken newKeyword)
		{
			return Update(newKeyword, AttributeLists, Type, ArgumentList, Initializer);
		}

		internal override SyntaxList<AttributeListSyntax> GetAttributeListsCore()
		{
			return AttributeLists;
		}

		internal override NewExpressionSyntax WithAttributeListsCore(SyntaxList<AttributeListSyntax> attributeLists)
		{
			return WithAttributeLists(attributeLists);
		}

		public new ObjectCreationExpressionSyntax WithAttributeLists(SyntaxList<AttributeListSyntax> attributeLists)
		{
			return Update(NewKeyword, attributeLists, Type, ArgumentList, Initializer);
		}

		public new ObjectCreationExpressionSyntax AddAttributeLists(params AttributeListSyntax[] items)
		{
			return WithAttributeLists(AttributeLists.AddRange(items));
		}

		internal override NewExpressionSyntax AddAttributeListsCore(params AttributeListSyntax[] items)
		{
			return AddAttributeLists(items);
		}

		public ObjectCreationExpressionSyntax WithType(TypeSyntax type)
		{
			return Update(NewKeyword, AttributeLists, type, ArgumentList, Initializer);
		}

		public ObjectCreationExpressionSyntax WithArgumentList(ArgumentListSyntax argumentList)
		{
			return Update(NewKeyword, AttributeLists, Type, argumentList, Initializer);
		}

		public ObjectCreationExpressionSyntax AddArgumentListArguments(params ArgumentSyntax[] items)
		{
			ArgumentListSyntax argumentListSyntax = ((ArgumentList != null) ? ArgumentList : SyntaxFactory.ArgumentList());
			return WithArgumentList(argumentListSyntax.AddArguments(items));
		}

		public ObjectCreationExpressionSyntax WithInitializer(ObjectCreationInitializerSyntax initializer)
		{
			return Update(NewKeyword, AttributeLists, Type, ArgumentList, initializer);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				1 => _attributeLists, 
				2 => _type, 
				3 => _argumentList, 
				4 => _initializer, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				1 => GetRed(ref _attributeLists, 1), 
				2 => Type, 
				3 => ArgumentList, 
				4 => Initializer, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitObjectCreationExpression(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitObjectCreationExpression(this);
		}

		public ObjectCreationExpressionSyntax Update(SyntaxToken newKeyword, SyntaxList<AttributeListSyntax> attributeLists, TypeSyntax type, ArgumentListSyntax argumentList, ObjectCreationInitializerSyntax initializer)
		{
			if (newKeyword != NewKeyword || attributeLists != AttributeLists || type != Type || argumentList != ArgumentList || initializer != Initializer)
			{
				ObjectCreationExpressionSyntax objectCreationExpressionSyntax = SyntaxFactory.ObjectCreationExpression(newKeyword, attributeLists, type, argumentList, initializer);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(objectCreationExpressionSyntax, annotations);
				}
				return objectCreationExpressionSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/ObjectCreationExpressionSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class ObjectCreationExpressionSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
		Friend _type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax

		Friend _argumentList As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax

		Friend _initializer As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax

		Public ReadOnly Property ArgumentList As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax)(Me._argumentList, 3)
			End Get
		End Property

		Public Shadows ReadOnly Property AttributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Get
				Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)(MyBase.GetRed(Me._attributeLists, 1))
			End Get
		End Property

		Public ReadOnly Property Initializer As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax)(Me._initializer, 4)
			End Get
		End Property

		Public Shadows ReadOnly Property NewKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationExpressionSyntax)._newKeyword, MyBase.Position, 0)
			End Get
		End Property

		Public ReadOnly Property Type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax)(Me._type, 2)
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal newKeyword As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax, ByVal attributeLists As SyntaxNode, ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax, ByVal argumentList As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax, ByVal initializer As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationExpressionSyntax(kind, errors, annotations, newKeyword, If(attributeLists IsNot Nothing, attributeLists.Green, Nothing), DirectCast(type.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeSyntax), If(argumentList IsNot Nothing, DirectCast(argumentList.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArgumentListSyntax), Nothing), If(initializer IsNot Nothing, DirectCast(initializer.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationInitializerSyntax), Nothing)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitObjectCreationExpression(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitObjectCreationExpression(Me)
		End Sub

		Public Function AddArgumentListArguments(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.WithArgumentList(If(Me.ArgumentList IsNot Nothing, Me.ArgumentList, Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ArgumentList()).AddArguments(items))
		End Function

		Public Shadows Function AddAttributeLists(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.WithAttributeLists(Me.AttributeLists.AddRange(items))
		End Function

		Friend Overrides Function AddAttributeListsCore(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.AddAttributeLists(items)
		End Function

		Friend Overrides Function GetAttributeListsCore() As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Return Me.AttributeLists
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Select Case i
				Case 1
					syntaxNode = Me._attributeLists
					Exit Select
				Case 2
					syntaxNode = Me._type
					Exit Select
				Case 3
					syntaxNode = Me._argumentList
					Exit Select
				Case 4
					syntaxNode = Me._initializer
					Exit Select
				Case Else
					syntaxNode = Nothing
					Exit Select
			End Select
			Return syntaxNode
		End Function

		Friend Overrides Function GetNewKeywordCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return Me.NewKeyword
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim red As SyntaxNode
			Select Case i
				Case 1
					red = MyBase.GetRed(Me._attributeLists, 1)
					Exit Select
				Case 2
					red = Me.Type
					Exit Select
				Case 3
					red = Me.ArgumentList
					Exit Select
				Case 4
					red = Me.Initializer
					Exit Select
				Case Else
					red = Nothing
					Exit Select
			End Select
			Return red
		End Function

		Public Function Update(ByVal newKeyword As Microsoft.CodeAnalysis.SyntaxToken, ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax), ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax, ByVal argumentList As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax, ByVal initializer As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Dim objectCreationExpressionSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			If (newKeyword <> Me.NewKeyword OrElse attributeLists <> Me.AttributeLists OrElse type <> Me.Type OrElse argumentList <> Me.ArgumentList OrElse initializer <> Me.Initializer) Then
				Dim objectCreationExpressionSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ObjectCreationExpression(newKeyword, attributeLists, type, argumentList, initializer)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				objectCreationExpressionSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, objectCreationExpressionSyntax1, objectCreationExpressionSyntax1.WithAnnotations(annotations))
			Else
				objectCreationExpressionSyntax = Me
			End If
			Return objectCreationExpressionSyntax
		End Function

		Public Function WithArgumentList(ByVal argumentList As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArgumentListSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.Update(Me.NewKeyword, Me.AttributeLists, Me.Type, argumentList, Me.Initializer)
		End Function

		Public Shadows Function WithAttributeLists(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.Update(Me.NewKeyword, attributeLists, Me.Type, Me.ArgumentList, Me.Initializer)
		End Function

		Friend Overrides Function WithAttributeListsCore(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.WithAttributeLists(attributeLists)
		End Function

		Public Function WithInitializer(ByVal initializer As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationInitializerSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.Update(Me.NewKeyword, Me.AttributeLists, Me.Type, Me.ArgumentList, initializer)
		End Function

		Public Shadows Function WithNewKeyword(ByVal newKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.Update(newKeyword, Me.AttributeLists, Me.Type, Me.ArgumentList, Me.Initializer)
		End Function

		Friend Overrides Function WithNewKeywordCore(ByVal newKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.WithNewKeyword(newKeyword)
		End Function

		Public Function WithType(ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectCreationExpressionSyntax
			Return Me.Update(Me.NewKeyword, Me.AttributeLists, type, Me.ArgumentList, Me.Initializer)
		End Function
	End Class
End Namespace